backend/shard/src/aggregator.rs

```diff
@@ -0,0 +1,229 @@
+        let tx_from_connection = tx_to_aggregator.clone().with(|msg| {
+            future::ok::<_,mpsc::SendError>(match msg {
+                Message::Connected => ToAggregator::ConnectedToTelemetryCore,
+                Message::Disconnected => ToAggregator::DisconnectedFromTelemetryCore,
+                Message::Data(data) => ToAggregator::FromTelemetryCore(data)
+            })
+        });
+
+        // Establish a resiliant connection to the core (this retries as needed):
+        let tx_to_telemetry_core = create_ws_connection(
+            tx_from_connection,
+            telemetry_uri
+        ).await;
+
+        // Handle any incoming messages in our handler loop:
+        tokio::spawn(Aggregator::handle_messages(rx_from_external, tx_to_telemetry_core));
+
+        // Return a handle to our aggregator:
+        Ok(Aggregator(Arc::new(AggregatorInternal {
+            conn_id: AtomicU64::new(1),
+            tx_to_aggregator,
+        })))
+    }
+
+    // This is spawned into a separate task and handles any messages coming
+    // in to the aggregator. If nobody is tolding the tx side of the channel
+    // any more, this task will gracefully end.
+    async fn handle_messages(mut rx_from_external: mpsc::Receiver<ToAggregator>, mut tx_to_telemetry_core: mpsc::Sender<FromAggregator>) {
+        use internal_messages::{ FromShardAggregator, FromTelemetryCore };
+
+        // Just as an optimisation, we can keep track of whether we're connected to the backend
+        // or not, and ignore incoming messages while we aren't.
+        let mut connected_to_telemetry_core = false;
+
+        // A list of close channels for the current connections. Send an empty tuple to
+        // these to ask the connections to be closed.
+        let mut close_connections: Vec<mpsc::Sender<()>> = vec![];
+
+        // Maintain mappings from the connection ID and node message ID to the "local ID" which we
+        // broadcast to the telemetry core.
+        let mut to_local_id = AssignId::new();
+
+        // Any messages coming from nodes that have been muted are ignored:
+        let mut muted: HashSet<LocalId> = HashSet::new();
+
+        // Now, loop and receive messages to handle.
+        while let Some(msg) = rx_from_external.next().await {
+            match msg {
+                ToAggregator::ConnectedToTelemetryCore => {
+                    // Take hold of the connection closers and run them all.
+                    let closers = close_connections;
+
+                    for mut closer in closers {
+                        // if this fails, it probably means the connection has died already anyway.
+                        let _ = closer.send(());
+                    }
+
+                    // We've told everything to disconnect. Now, reset our state:
+                    close_connections = vec![];
+                    to_local_id.clear();
+                    muted.clear();
+
+                    connected_to_telemetry_core = true;
```

Comment:
It looks to me like `connected_to_telemetry_core` should be more of an enum like:

```rust
enum State {
    Connected {
        close_connections: ...,
        to_local_id: ...,
        muted: ...,
    },
    Disconnected,
}
```
So that whoever touches the code later cannot introduce a bug that puts the aggregator into an invalid state (e.g.: read or write to `muted` without checking for connection status first).

Reply:
Ooh, I like that, thanks!

Reply:
Alas it's not quite that straightforward, because I want to hold onto the `close_connections` while we are disconnected, and call them when we reconnect. But I'll keep this in mind; something very similar might work well!